rspace/src/main/scala/coop/rchain/rspace/spaces/FineGrainedRSpace.scala

```diff
@@ -161,137 +169,139 @@ class FineGrainedRSpace[F[_], C, P, E, A, R, K] private[rspace] (
   override def produce(channel: C, data: A, persist: Boolean, sequenceNumber: Int)(
       implicit m: Match[P, E, A, R]
   ): F[Either[E, Option[(ContResult[C, P, K], Seq[Result[R]])]]] =
-    syncF.delay {
-      Kamon.withSpan(produceSpan.start(), finishSpan = true) {
-        val span = Kamon.currentSpan()
-        span.mark("before-produce-ref-computed")
-        val produceRef = Produce.create(channel, data, persist, sequenceNumber)
-        span.mark("before-produce-lock")
-        produceLock(channel) {
-          span.mark("produce-lock-acquired")
-          //TODO fix double join fetch
-          val groupedChannels: Seq[Seq[C]] = store.withTxn(store.createTxnRead()) { txn =>
-            store.getJoin(txn, channel)
-          }
-          span.mark("grouped-channels")
-          logger.debug(s"""|produce: searching for matching continuations
+    contextShift.evalOn(scheduler) {
```

Comment:
maybe a similar refactor to replay (lockedConsume, lockedProduce)?

Reply:
I don't see a point at the moment